src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -86,33 +88,56 @@ public class HubConnectionContext
 
         public int? LocalPort => Features.Get<IHttpConnectionFeature>()?.LocalPort;
 
-        public async Task WriteAsync(HubMessage message, bool throwOnFailure = false)
+        public virtual async Task WriteAsync(HubMessage message)
         {
-            while (await _output.Writer.WaitToWriteAsync())
+            var buffer = ProtocolReaderWriter.WriteMessage(message);
+
+            // Hubs support multiple producers so need to lock as we write into the buffer
+            // since the pipe only supports a single consumer at a time
+            lock (_writeLock)
             {
-                if (_output.Writer.TryWrite(message))
-                {
-                    return;
-                }
+                _connectionContext.Transport.Output.Write(buffer);
             }
 
-            _logger.OutboundChannelClosed();
+            Interlocked.Exchange(ref _lastSendTimestamp, Stopwatch.GetTimestamp());
 
-            if (throwOnFailure)
+            await FlushAsync(_connectionContext.Transport.Output, CancellationToken.None);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private Task FlushAsync(PipeWriter pipeWriter, CancellationToken cancellationToken)
+        {
+            var awaitable = pipeWriter.FlushAsync(cancellationToken);
+            if (awaitable.IsCompleted)
             {
-                throw new OperationCanceledException("Outbound channel was closed while trying to write hub message");
+                // The flush task can't fail today
+                return Task.CompletedTask;
             }
+            return FlushAsyncAwaited(awaitable, cancellationToken);
         }
 
-        public async Task DisposeAsync()
+        private async Task FlushAsyncAwaited(ValueAwaiter<FlushResult> awaitable, CancellationToken cancellationToken)
         {
-            // Nothing should be writing to the HubConnectionContext
-            _output.Writer.TryComplete();
+            // Since the flush awaitable doesn't currently support multiple awaiters
+            // we need to use a task to track the callbacks.
+            // All awaiters get the same task
+            lock (_flushLock)
+            {
+                if (_flushTcs == null || _flushTcs.Task.IsCompleted)
```

Comment:
If the TCS isn't null or complete, we just tack another wait on it, but we have a different awaitable than the one that the `_flushTcs` is tracking. If I call our `FlushAsync` twice simultaneously doesn't that mean I call the Pipe `FlushAsync` twice but both tasks only end up waiting on the first one?

Reply:
Yes, thats by design. The there's only a single thing to wait on and that's backpressure was released. You don't get a task for your individual write.

Reply:
Ok, is that what L123 is trying to say? It wasn't entirely clear to me at first reading.

Reply:
> but we have a different awaitable than the one that the _flushTcs is tracking.

This might be where there's a misunderstanding? We're ignoring the returned awaitable while the flush task is incomplete. So it looks like,

FlushAsync() -> complete
FlushAsync() -> complete
FlushAsync() -> back pressure (waiting on flush tcs ignoring the returned awaitable)
FlushAsync() -> back pressure (waiting on flush tcs ignoring the returned awaitable)
FlushAsync() -> back pressure (waiting on flush tcs ignoring the returned awaitable)

Since the pipe is single producer single consumer, the awaitable returned represents the state of the PipeWriter itself, not the individual write operation. You can't have multiple callers waiting on the same FlushAsync call, as such, it isn't even possible to FlushAsync twice because the second one will throw.